### ISSUE TITLE
Video for checking token ownership (CEP-78)

### ISSUE THREAD

**Issue by ipopescu** [casper-network/docs#872]
After the first [NFT migration video](https://www.youtube.com/watch?v=Co3VdPypc_o), we need to show how someone can check token ownership. The content to be covered in this second video is all available [here](https://github.com/casper-ecosystem/cep-78-enhanced-nft/blob/dev/tutorials/token-ownership-tutorial.md).

Requirements for the second video:
- Create a script and slides based on existing tutorials (2 points)
- Record and edit (5 points)
- Add additional details for publishing: description, tags, links to specific points in the video (0-1 point)
- Work with Nico from the Association to add a description, tags, and to publish it
   -  This video needs to be in the same playlist as the first [NFT migration video](https://www.youtube.com/watch?v=Co3VdPypc_o)

**Comment by RitaMAllenCA** [comment]
waiting on branding resources from the association, currently unblocked.

**Comment by ShalmaliCL** [comment]
Thank you for the update, @ipopescu. Will make sure the script of this video is in sync with the first [NFT migration video](https://www.youtube.com/watch?v=Co3VdPypc_o).

**Comment by ShalmaliCL** [comment]
Creating the required material [here.](https://drive.google.com/drive/folders/1vk-gREdft9YGetSGJPjtoZFkTCYoUmnf)

**Comment by ShalmaliCL** [comment]
The CEP-78 token ownership tutorial video is ready.
https://drive.google.com/file/d/1dlyTqKzreJgE_bRqMNUjj4AzTc4h3ikw/view?usp=sharing
Reviewers: @ACStoneCL @ramujin

**Comment by ipopescu** [comment]
Thanks, Shalmali, for the video! I completed the review and left notes in the PPT presentation. Also, for the video I have the following feedback:

- IMO, the discussion on trust needs to be much more emphasized. I suggest adding this based on what we already have in the docs.
- When doing a screen capture, try to hide other tabs in your browser.
- Min 1:08 there is some background noise - perhaps it can be edited?
- Min 4:32 visual information on the token ID (a slide or showing an NFT token where the token ID is used) would be beneficial. - Min 4:45 it would be good to actually show how the token owner would be set to Hash upon token creation.
- It would be useful to demo how to navigate to the hash_by_index dictionary using the block explorer and to run the query shown in the slides and the documentation.
- Min 6:50 Explain that the page_table uRef is used as the seed uRef in the command. 
- Min 7:40 Seems to be an abrupt transition. Perhaps moving the audio explaining that we will use the page_0 dictionary to get token information would help the transition. Then you can show the query.
- 8:02 Highlighting and mentioning the page_table uref is confusing. Perhaps it would be better to mention and show the page_0 uref directly (I see you are using uref-838…).

**Comment by ramujin** [comment]
I agree with Iulia's comments and I also left you some comments in the PPT file. Additionally, I think you should have also circulated the script (perhaps you did?) for review before recording so we could have provided feedback before the large time investment on your part to film and edit this content.

Here are a few more points:

* On the first slide, you have "e" bolded by itself in the word "the"
* 1:08 – agreed, the volume change at least should be addressed because it suddenly gets louder
* 2:45 – the volume gets soft again
* 2:55 – you should show what you mean when you say "I have already installed CEP-78 contract"
* 3:50 – you're talking about a named key CEP-78, but you're still just showing the boolean values
* 4:00 – you should explain what you mean by "page zero from the page table dictionary"
* 4:05 – "Considering that the token number is token id, this account owns token 1" – that is not readily apparent at all
* 5:20 – what would they do with the returned hash? it isn't really tying to ownership
* 5:50 – why is the contract-centric approach more secure? why would I not trust my account's keys? a little clarification would help a ton here
* 6:00 – it's unclear what makes this a contract-centric approach when it still requires the account hash
* This video leaves me with quite a few questions:
  * It seems like the page table could use a separate discussion on its own... why is it being used at all?
  * Do I always have to first query the page table before querying subsequent tables?
  * Is this the default configuration or do I have to customize the NFT to get this behavior?
  * I am still unclear on the trust/security points being raised. They seem very important to know but are just mentioned as passing points.
  * This is more of a general question to the team, but is there no simpler way to just see who owns a token? This seems like a serious engineering oversight to make a user jump through hoops just to query a simple property. I imagine this would be a frequent operation for NFT developers.

**Comment by ShalmaliCL** [comment]
@ramujin
Thank you for your review.
I will look into the suggestions and incorporate them. However, I did get the script reviewed by you and Adam before recording it.
@ipopescu 
Thank you for the review. However, I am waiting for your script review to start looking into the suggestions.

**Comment by ramujin** [comment]
Ok great! That’s why I had posed the question. I couldn’t recall if you did.

**Comment by ipopescu** [comment]
Thanks, @ShalmaliCL. I have reviewed the script. Please combine the comments I left open in the PPT with the comments I left in the script. Thank you!

**Comment by ipopescu** [comment]
@ramujin I have heard similar feedback from other developers, and I think there were several feedback sessions regarding your point: "This is more of a general question to the team, but is there no simpler way to just see who owns a token?...".

**Comment by ShalmaliCL** [comment]
@ramujin @ipopescu - Thank you for the review. I will work on the suggestions and get back to you.

**Comment by ShalmaliCL** [comment]
I have incorporated the above suggestions into [the presentation ](https://docs.google.com/presentation/d/1_a4RiLSzSbAXlTceUqWMWAkprchezVdC/edit?usp=sharing&ouid=103877774796379801362&rtpof=true&sd=true) and [the script](https://docs.google.com/document/d/13YcdlDCd0t1hJLINMpPn2MOVX6jPrDZd0_xbEUUJZmE/edit?usp=sharing).
@ipopescu @ramujin @ACStoneCL - Request you to review and advise.

**Comment by ipopescu** [comment]
@ShalmaliCL, please address all the points already discussed before asking for a re-review. This saves us review cycles and time.
Also, note that you need to incorporate [PR 222](https://github.com/casper-ecosystem/cep-78-enhanced-nft/pull/222) as we discussed yesterday, before the review request. Here is my current feedback.

The script:
- The discussion on trust is still missing from the first round of reviews.
- You need to include the updates I made in PR 222 (https://github.com/casper-ecosystem/cep-78-enhanced-nft/pull/222).

The presentation:
- The FAQ section needs more visual organization and details; another option is to talk about the FAQs without slides and show your account instead.
- The logo is still missing from the first round.
- I’ve made some formatting updates to address open comments from the first round.
- You need to include the updates I made in PR 222 (https://github.com/casper-ecosystem/cep-78-enhanced-nft/pull/222).

**Comment by ShalmaliCL** [comment]
@ipopescu - Thank you for your review.
I have made the following changes to the script and the presentation. Also, please find the response to the inputs suggested by you:

**The script:**
< Iulia >"The discussion on trust is still missing from the first round of reviews." 

< Shalmali > As discussed yesterday points related to trust were already mentioned in the script but later in the flow/script. Now I have moved the text related to trust to the beginning of the script

< Iulia >"You need to include the updates I made in PR 222 (https://github.com/casper-ecosystem/cep-78-enhanced-nft/pull/222)." 

< Shalmali > Before sending this for review I had already done the following changes:
- the two methods diagram
- OwnerReverseLookupMode prerequisite
- A note about the owner of a token when identity mode was set to hash
- There were a few points, where I thought it is okay to have a different view than [this tutorial](https://github.com/casper-ecosystem/cep-78-enhanced-nft/blob/dev/tutorials/token-ownership-tutorial.md). Example: I used, "Being a Casper user you may want to understand" instead of "As someone interacting with an NFT contract". Now I have used the script that you have suggested.

The presentation:

I have included The FAQ in the demo script. 
< Iulia > The logo is still missing from the first round. 
< Shalmali > I was not sure which logo to use as I did use a presentation template from the "Upgrading a CEP-78 Casper NFT" tutorial. Now, understood the logo from [this video](https://www.youtube.com/watch?v=Co3VdPypc_o), searched into our Casper drive and updated a similar logo into the presentation. 

< Iulia >You need to include the updates I made in PR 222 (https://github.com/casper-ecosystem/cep-78-enhanced-nft/pull/222)
< Shalmali > 

- I have updated the prerequisite slide. 
- Moved FAQs to demo script. 
- However, the diagram which is part of the PR was already updated in the presentation before I sent it for review.
- Complete notes were part of the "Transcript of CEP-78_Token_Ownership", now I have added the same notes in the presentation as well after our discussion yesterday.

After our huddle call yesterday, and further to our agreement on the flow and the changes, I incorporated all the suggestions. I also sent a request to re-review on Slack yesterday just to speed up the process.

Do let me know if you have any further inputs and will be happy to incorporate them. If required, to save time and finish the task quickly, I can schedule a call with you and the team. 
Thank you!

**Comment by ipopescu** [comment]
@ShalmaliCL, thanks. I left some further minor feedback. Please fix the explanation for FAQ #1 about the page table.

**Comment by ShalmaliCL** [comment]
Hi @ipopescu,
I have incorporated your suggestions. Still, there are 3 open comments. 2 are for Ramsin to approve and 1 comment needs your approval/help. I have tagged you and Ramsin too to the respective comments. 

Hi @ramujin,
I will be re-recording the video anyway and I will keep note of your suggestions while recording it. Please find my responses inline and let me know what you think:

< Ramsin > This video leaves me with quite a few questions:
< Ramsin > 
- It seems like the page table could use a separate discussion on its own... why is it being used at all? 
- Do I always have to first query the page table before querying subsequent tables? 
@ipopescu  - Could you please help to answer these questions?

< Ramsin > Is this the default configuration or do I have to customize the NFT to get this behaviour?
< Shalmali > I have added the text related to this and tagged you for approval. Let me know what you think.

< Ramsin > I am still unclear on the trust/security points being raised. They seem very important to know but are just mentioned as passing points.
< Shalmali > With Iulia's help, I have added notes and points related to trust in at least 4 different places. Hope this will solve the issue.

**Comment by ipopescu** [comment]
@ShalmaliCL, I completed another review and left comments regarding technical accuracy, now that the FAQ was moved to another method. Please take a look and hold off on recording until the script is finalized.

**Comment by ramujin** [comment]
@ShalmaliCL I went through another pass and answered your questions as well. My general feedback is that it would be much more helpful to contextualize things for the developer when going through each step of the process, answering questions about why something is being done and the importance of elements as you are using them. Hopefully, the comments I left highlight what I'm talking about.

**Comment by ShalmaliCL** [comment]
@ramujin @ACStoneCL @ipopescu 
I have shared the recorded video for review over Slack. Once the video is final, I will upload it to google drive to share further.
Request you to review and let me know what you think.

**Comment by ramujin** [comment]
@ShalmaliCL I have done another review, and this video is much better! Great work. I still have some questions and feedback for you:

* 0:20 – Abruptly mentioning something about setting some modality to complete without any further explanations (it’s covered later, so I suggest just not mentioning it here)
* 1:12 – Awkward long pause
* 1:45 – For the prerequisites, you should point the developers to the resources to learn them if they do not yet know them
* 5:10-6:00 – You spend almost a minute discussing things while showing just a title slide. It would be better if you show a slide with bullet points highlighting what you’re talking about
* 8:00 – Why is this method more secure? What is the security risk?
* 13:23 – Awkward long pause again
* 14:00 – Why not circumvent this whole process and directly work with the token_owners dictionary from the beginning? Does that not get me the same result since I have to iterate over the page table dictionary data anyway?

**Comment by ipopescu** [comment]
@ShalmaliCL @ramujin, here are a few additional points, with some comments added directly in the transcript.

1. 0:20 - I originally asked Shalmali to mention this here, but since it is already mentioned in the prerequisites later, it can be edited and removed.
2. For additional details regarding trust, you can review the Zoom recording sent on Slack, especially the summary at min 39. The CEP-78 contract is the final source of truth.
3. See the token_owner discussion at min 30 of the Zoom recording.
   - Token owners: mapping of token IDs to token owners. If you want to know who owns token X, you use the token owners dictionary.
   - In this tutorial, we want to figure out the opposite (and map an account to a token ID)… which tokens does X own?

**Comment by ShalmaliCL** [comment]
@ipopescu @ramujin - Updated script and the video with the following feedback points:
- [ ] 0:20 – **Removed modality point** 
- [ ] 1:12 – **Awkward long pause removed**
- [ ] 1:45 – **For the prerequisites, the link will be added to the video description**
- [ ] 5:10-6:00 – _You spend almost a minute discussing things while showing just a title slide. It would be better if you show a slide with bullet points highlighting what you’re talking about_ **Added more text to the slide.**

- [ ] 8:00 – _Why is this method more secure? What is the security risk?_ **Added more about security** 
- [ ] 13:23 – _Awkward long pause again_ **Removed**

I have shared the video over slack, please let me know what you think. Thank you!

**Comment by ipopescu** [comment]
Reviewed, LGTM!

**Comment by ShalmaliCL** [comment]
Reviewed and the final video is [here](https://drive.google.com/drive/folders/1vk-gREdft9YGetSGJPjtoZFkTCYoUmnf?usp=sharing).
